### PR TITLE
CORS-3931: [azure] add more permission when installing cluster in existing vnet

### PR DESCRIPTION
**Pull request by jinyunma**
1. Panic error is thrown from getNextAvailableIP when installing in shared VPC with minimal permissions, see failed job: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9422/pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc/1885421707863265280 ([slack thread](https://redhat-internal.slack.com/archives/C01V1DP387R/p1738888488097589))

Add additional permission "Microsoft.Network/virtualNetworks/checkIpAddressAvailability/read" to fix the issue

2. Read install-config and assign proper permission list based on various config, remove env ENABLE_MIN_PERMISSION_FOR_MARKETPLACE and ENABLE_MIN_PERMISSION_FOR_DES.
3.  Extract custom role creation from step "azure-provision-service-principal-minimal-permission" as a separated step "azure-provision-custom-role", to cover case "install cluster with managed identity assigned minimal permissions" which only needs to custom role. 
And introduce new chain "azure-provision-service-principal-minimal-permission" to include both steps.
4. Extract role assignment from step `azure-provision-disk-encryption-set`, `azure-provision-vnet`,`ipi-conf-azure-resourcegroup` as separated steps, only run role assignment steps in dedicated jobs instead of using ENV to control it.

### CODE REVIEW

**Comment by jinyunma** [comment]
/pj-rehearse pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc

**Comment by jinyunma** [comment]
Installation on shared vpc succeeded in [job](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/61374/rehearse-61374-pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc/1887721408583176192)

/pj-rehearse ack

cc @jianlinliu @gpei to review

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 1]
```diff
@@ -252,6 +252,13 @@ required_permissions="""
 \"Microsoft.Storage/storageAccounts/listKeys/action\"
 """
 
+# optional permissions when installing cluster in existing vnet
+required_permissions="""
+\"Microsoft.Network/virtualNetworks/checkIpAddressAvailability/read\",
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:257]: The permission is needed by recent PR change, but not for all the previous OCP versions, right?

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:257]: should be related with bug https://issues.redhat.com//browse/OCPBUGS-37442, which has been backported to 4.17. So this permission is required on 4.17+ CAPI install. 
Will file doc bug to request to add the permission on 4.17+.

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:257]: Error happens when creating manifests, this should impact both IPI and UPI. 
But I can update to only apply permission on 4.17+.

Comment (gpei) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh]: @jinyunma one more question, does it only affect the installation for existing vnet? If so, do we need a separate env var to control it, so we can skip this permission for IPI install without existing vnet?

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh]: Good suggestion.
As discussed offline, will read install-config and assign proper permissions based on various config.

**Comment by jinyunma** [comment]
/pj-rehearse pull-ci-openshift-installer-main-azure-ovn-marketplace-images periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-fips-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-amd64-nightly-azure-ipi-marketplace-mini-perm-f7

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml [hunk 2]
```diff
@@ -14,6 +12,8 @@ chain:
   - ref: ipi-install-rbac
   - ref: openshift-cluster-bot-rbac
   - ref: ipi-install-hosted-loki
+  - ref: azure-provision-service-principal-minimal-permission
+  - ref: azure-provision-bastionhost-managed-identity
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml:16]: Seem like `azure-provision-bastionhost-managed-identity` does not depend on `azure-provision-service-principal-minimal-permission`, while  `ipi-conf-azure-oidc-creds-provision` depends on `azure-provision-service-principal-minimal-permission` when `ENABLE_MIN_PERMISSION_FOR_STS` enabled, so I think `azure-provision-service-principal-minimal-permission` need to be prior to `ipi-conf-azure-oidc-creds-provision`

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml:16]: The issue also happened on other oidc + mini-perm install workflow.

Comment (jinyunma) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml]: good catch, I moved step `azure-provision-service-principal-minimal-permission` before step `ipi-conf-azure-oidc-creds-provision`.

> Seem like azure-provision-bastionhost-managed-identity does not depend on azure-provision-service-principal-minimal-permission

yes, there is no dependency now, since there is some issues with managed identity assigned minimal permissions, need to debug further and update step later to support it.

**Comment by jinyunma** [comment]
/pj-rehearse pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-fips-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml [hunk 3]
```diff
@@ -4,11 +4,11 @@ chain:
...
   - chain: azure-provision-bastionhost
-  - ref: azure-provision-service-principal-minimal-permission
-  - ref: azure-provision-bastionhost-managed-identity
   - chain: ipi-conf-azure
   - ref: ipi-conf-azure-sharednetwork
   - ref: ipi-conf-manual-creds
+  - ref: azure-provision-service-principal-minimal-permission
+  - ref: azure-provision-bastionhost-managed-identity
   - ref: ipi-conf-azure-oidc-creds-provision
   - ref: ipi-conf-azure-provisioned-resourcegroup
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml:13]: If possible, let us move `ipi-conf-azure-provisioned-resourcegroup` a bit before `azure-provision-service-principal-minimal-permission` to be safer.

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/cco-manual-workload-identity/managed-identity/provision/cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity-managed-identity-provision-chain.yaml:13]: Or ignore my comments, `ipi-conf-azure-provisioned-resourcegroup` depends on `ipi-conf-azure-oidc-creds-provision`, so no need any update.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 4]
```diff
@@ -136,282 +104,39 @@ fi
...
+    fi
+    cat <<EOF > "${SHARED_DIR}/${os_sp_file_name}"
+{"subscriptionId":"${AZURE_AUTH_SUBSCRIPTOIN_ID}","clientId":"${sp_id}","tenantId":"${sp_tenant}","clientSecret":"${sp_password}"}
 EOF
+
     # for destroy
-    echo "${sts_sp_id}" >> "${SHARED_DIR}/azure_sp_id"
-    rm -rf ${sts_sp_output}
-fi
+    echo "${sp_id}" > "${SHARED_DIR}/azure_sp_id"
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:136]:
```suggestion
    echo "${sp_id}" >> "${SHARED_DIR}/azure_sp_id"
```

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 5]
```diff
@@ -136,282 +104,39 @@ fi
...
+    rm -f ${sp_output}
+
+    # ensure that role assignment creation is successful
+    cmd="az role assignment list --role {role_name}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:140]: This step is used to create sp, no actions related to `az role assignment`, do we need to check it here ?

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:140]: Here create sp with customer role, and need sometime to waiting for role assignment successful, else following steps might fail, e.g step `upi-install-azure`, so add this check to ensure that role assignment succeeded.

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-ref.yaml [hunk 6]
```diff
@@ -19,18 +19,12 @@ ref:
...
-    default: "false"
+  - name: AZURE_MINIMAL_PERMISSIONS_USAGE
+    default: "cluster"
     documentation: |-
-      Assign addtional permissions for SP to install cluster with marketplace image
-  - name: ENABLE_MIN_PERMISSION_FOR_DES
-    default: "false"
-    documentation: |-
-      Assign additional permissions for SP to install cluster configured disk encryption set
-  - name: ENABLE_MIN_PERMISSION_FOR_STS
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-ref.yaml:30]: Maybe we can keep this ENV, or else, we need to update many places beside job config yaml, also `ipi-conf-azure-oidc-creds-deprovision` and `ipi-conf-azure-oidc-creds-provision`. This also help keep consistent with gcp. AWS is using `AWS_CCOCTL_USE_MINIMAL_PERMISSIONS` ENV var to control. Or else, we would have 3 different ENV var to control the behavior.

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 7]
```diff
@@ -41,43 +41,27 @@ function run_cmd_with_retries_save_output()
...
-    cmd="az role definition create --role-definition ${role_definition}"
-    run_command "${cmd}" || return 1
+    local cmd="$1" retries="${2:-}"
+    local try=0 ret=0
+    [[ -z ${retries} ]] && max="20" || max=${retries}
+    echo "Trying ${max} times max to run '${cmd}'"
 
-    echo "Sleep 1 min to wait for custom role created"
-    sleep 60
+    res=$(eval "${cmd}") || ret=$?
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:51]:
```suggestion
    eval "${cmd}" || ret=$?
```

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 8]
```diff
@@ -41,43 +41,27 @@ function run_cmd_with_retries_save_output()
...
+    while [[ ${ret} -ne 0 || -z "${res}" ]] && [ ${try} -lt ${max} ]; do
+        echo "'${cmd}' did not return success or return empty, waiting 60 sec....."
+        sleep 60
+        try=$(( try + 1 ))
+        ret=0
+        res=$(eval "${cmd}") || ret=$?
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:57]:
```suggestion
        eval "${cmd}" || ret=$?
```

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh (hunk 7 above)
Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:51]: ${cmd} might execute and give empty output, but return code is still 0, so here save output to check. The successful result is that output is not empty and return code is 0.

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-ref.yaml (hunk 6 above)
Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-ref.yaml:30]: If keep env `ENABLE_MIN_PERMISSION_FOR_STS`, maybe need to introduce new env to control cluster sp creation, any thoughts?

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

File: ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh [hunk 9]
```diff
@@ -136,282 +104,42 @@ fi
...
-
-    create_sp_with_custom_role "${sts_sp_name}" "${sts_sp_role}" "${AZURE_AUTH_SUBSCRIPTOIN_ID}" "${sts_sp_output}"
-    sts_sp_id=$(jq -r .appId "${sts_sp_output}")
-    sts_sp_password=$(jq -r .password "${sts_sp_output}")
-    sts_sp_tenant=$(jq -r .tenant "${sts_sp_output}")
+sp_list=""
+[[ "${AZURE_SP_CREATION_FOR_CLUSTER}" == "true" ]] && sp_list="${sp_list} cluster"
+[[ "${ENABLE_MIN_PERMISSION_FOR_STS}" == "true" ]] && sp_list="${sp_list} sts"
+for sp_type in ${sp_list}; do
+    role_name=$(jq -r ".${sp_type}" "${SHARED_DIR}/azure_custom_role_name")
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh:112]: In `cucushift-installer-rehearse-azure-ipi-private-networking-type-minimal-permission` case, there is no `azure_custom_role_name` generated, right?

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh]: right, fixed, moved this line into line114 `if` block.

File: ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+    # create custom role
+    cmd="az role definition create --role-definition ${role_definition}"
+    run_command "${cmd}" || return 1
+
+    echo "Sleep 1 min to wait for custom role created"
+    sleep 60
+}
+
+if [[ "${AZURE_INSTALL_USE_MINIMAL_PERMISSIONS}" == "no" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:82]: We do not need this block any more.

Comment (jianlinliu) [reply ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:82]: Or append `&& [[ "${ENABLE_MIN_PERMISSION_FOR_STS}" == "false" ]]`

File: ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    az cloud set --name AzureCloud
+fi
+az login --service-principal -u "${AZURE_AUTH_CLIENT_ID}" -p "${AZURE_AUTH_CLIENT_SECRET}" --tenant "${AZURE_AUTH_TENANT_ID}" --output none
+
+# create custom role with minimal permission for cluster to create the infrastructure.
+if [[ "${AZURE_INSTALL_USE_MINIMAL_PERMISSIONS}" == "no" ]]; then
+    echo "ENV AZURE_INSTALL_USE_MINIMAL_PERMISSIONS is set to no, skip to create custom role with minimal permission for cluster!"
+else
+
+    CLUSTER_NAME="${NAMESPACE}-${UNIQUE_HASH}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:131]: The following `ENABLE_MIN_PERMISSION_FOR_STS` block also need to reference this VAR.

File: ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+\"Microsoft.KeyVault/vaults/write\",
+\"Microsoft.KeyVault/vaults/delete\",
+\"Microsoft.KeyVault/vaults/deploy/action\",
+\"Microsoft.KeyVault/vaults/keys/read\",
+\"Microsoft.KeyVault/vaults/keys/write\",
+${required_permissions}
+"""
+    fi
+
+    custom_role_name_json="{}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:337]: The following `ENABLE_MIN_PERMISSION_FOR_STS` block also need to reference this var.

File: ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh (hunk 10 above)
Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:82]: append `&& [[ "${ENABLE_MIN_PERMISSION_FOR_STS}" == "false" ]] ` to print message that "skip the step when both envs are not set"

File: ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+# log in with az
+if [[ "${CLUSTER_TYPE}" == "azuremag" ]] || [[ "${CLUSTER_TYPE}" == "azurestack" ]]; then
+    echo "Installation with minimal permissions is only supported on Azure Public Cloud so far, exit..."
+    exit 1
+else
+    az cloud set --name AzureCloud
+fi
+az login --service-principal -u "${AZURE_AUTH_CLIENT_ID}" -p "${AZURE_AUTH_CLIENT_SECRET}" --tenant "${AZURE_AUTH_TENANT_ID}" --output none
+
+if [[ "${AZURE_INSTALL_USE_MINIMAL_PERMISSIONS}" == "no" ]] && [[ "${ENABLE_MIN_PERMISSION_FOR_STS}" == "false" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh:126]: Seem like it is placed in a wrong place, I meant line 82.

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/custom-role/azure-provision-custom-role-commands.sh]: yes, should update at line82, fixed.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc pull-ci-openshift-installer-main-azure-ovn-marketplace-images periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-fips-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-amd-f28-destructive

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7 periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-arm-f14

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7 periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7

**Comment by jinyunma** [comment]
Tested job:
[pull-ci-openshift-installer-main-e2e-azure-ovn-shared-vpc](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/61374/rehearse-61374-pull-ci-openshift-installer-main-azure-ovn-marketplace-images/1888402730209251328)
[pull-ci-openshift-installer-main-azure-ovn-marketplace-images](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/61374/rehearse-61374-pull-ci-openshift-installer-main-azure-ovn-marketplace-images/1888402730209251328)
[periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/61374/rehearse-61374-periodic-ci-openshift-verification-tests-master-installation-nightly-4.18-azure-ipi-default-mini-perm-arm-f7/1888227163082067968)
[periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/61374/rehearse-61374-periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-upi-mini-perm-amd-f28-destructive/1888402730301526016)
[periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-arm-f14](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/61374/rehearse-61374-periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-des-mini-perm-arm-f14/1888520205114544128)
[periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/61374/rehearse-61374-periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-oidc-mini-perm-arm-f7/1888791128795254784)
[periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/61374/rehearse-61374-periodic-ci-openshift-openshift-tests-private-release-4.18-multi-nightly-azure-ipi-private-spec-net-type-spec-perm-arm-f7/1888851770407391232)

@jianlinliu @gpei  PTAL again, thanks.

**Comment by jinyunma** [comment]
Rebase to solve conflict.

@jianlinliu @gpei PTAL, thanks!

File: ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+cluster_sp_id=${AZURE_AUTH_CLIENT_ID}
+if [[ -f "${SHARED_DIR}/azure_minimal_permission" ]]; then
+    cluster_sp_id=$(jq -r '.clientId' "${SHARED_DIR}/azure_minimal_permission")
+fi
+
+# Assign system role "Contributor" to cluster sp on scope of resource group where cluster to be created.
+if [[ -n "${cluster_rg}" ]]; then
+    cluster_rg_id=$(az group show -g "${cluster_rg}" --query id -otsv)
+    echo "Assigin role 'Contributor' to resource group ${cluster_rg}"
```
Comment (patrickdillon) [inline ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh:45]: nit:
```suggestion
    echo "Assign role 'Contributor' to resource group ${cluster_rg}"
```
Perhaps even better...
```suggestion
    echo "Assign role 'Contributor' to ${cluster_sp_id} with scope over resource group ${cluster_rg}"
```

File: ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+trap 'if [[ "$?" == 0 ]]; then EXIT_CODE=0; fi; echo "${EXIT_CODE}" > "${SHARED_DIR}/install-pre-config-status.txt"' EXIT TERM
+
+CONFIG="${SHARED_DIR}/install-config.yaml"
+
+cluster_rg=$(yq-go r ${CONFIG} 'platform.azure.resourceGroupName')
+vnet_rg=$(yq-go r ${CONFIG} 'platform.azure.networkResourceGroupName')
+
+if [[ -z "${cluster_rg}" ]] && [[ -z "${vnet_rg}" ]]; then
+    echo "This step used to grant proper permissions on scope of cluster rg or vnet rg, but both rg are empty, abort..."
+    exit 1
```
Comment (patrickdillon) [inline ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh:22]: hm should this be `exit 1` or `exit 0` (no error)?  I would have thought `exit 0` because there may be valid jobs that don't set resource group or particularly the network rg?

**Review by patrickdillon — COMMENTED** [review]
One question about the exit codes, and a nit... (don't bother fixing the nit unless you're making other changes).

Also I would be very interested in a longer term plan to get some of the azure permissions logic into the installer, both to benefit users and to facilitate docs. I'm not sure if there is anything in place right now for the docs team to help keep the perms in sync, or is it all just done manually (by you)?

File: ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh (hunk 15 above)
Comment (jinyunma) [reply ci-operator/step-registry/ipi/conf/azure/resourcegroup/assign-role/ipi-conf-azure-resourcegroup-assign-role-commands.sh]: Currently, this step only inserts into specific install workflow to cover one QE test case, I was thinking that is `exit 1` if both rgs are empty to make sure the step be performed in expected workflow.
But if it add into other valid jobs in the future on purpose, that make sense to be `exit 0` here, updated.

**Comment by jinyunma** [comment]
> I'm not sure if there is anything in place right now for the docs team to help keep the perms in sync, or is it all just done manually (by you)?

AFAIK, there is no place to keep the perms in sync so far, all are done manually. 
When detecting missing perms during the test, we update in prow steps and file a doc bug to request doc update.

**Comment by jinyunma** [comment]
/retitle CORS-3931: [azure] add more permission when installing cluster in existing vnet

**Comment by jianlinliu** [comment]
cc @liangxia to approve

File: ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml [hunk 16]
```diff
@@ -553,9 +553,9 @@ tests:
     post:
     - chain: ipi-azure-post
     pre:
-    - ref: azure-provision-service-principal-minimal-permission
     - ref: ipi-conf
     - ref: ipi-conf-azure-default
+    - chain: azure-provision-service-principal-minimal-permission
```
Comment (liangxia) [inline ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: Do we really want to change `openshift-priv` ones ?

Comment (jinyunma) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: Are those openshift-priv jobs automatically synced up by some way? I searched them when updating this chain.
I'm not sure when this job will be triggered. Once triggered, the change is needed here.

Comment (jianlinliu) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: Yeah, we updated the `openshift-priv` ones on purpose because of mini-perm change.

Comment (liangxia) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: AFAIK, those `openshift-priv` are used by ART team.  Generally we don't touch them as there are auto sync mechanism. 
If you are sure this is exactly needed, I can approve it. Just let me know.

Comment (jinyunma) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: got it, the change is required regardless in which way to update them.
Thanks for approving this.

Comment (jinyunma) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml]: @liangxia I removed the change in `openshift-priv` job config, PTAL.

Comment (jinyunma) [reply ci-operator/config/openshift-priv/installer/openshift-priv-installer-main.yaml:558]: seems that the change must be done manually in this PR for `openshift-priv` jobs, otherwise job [pull-ci-openshift-release-master-ci-operator-config](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/61374/pull-ci-openshift-release-master-ci-operator-config/1894257025961955328) failed, because some envs are removed on purpose in this PR.

So I changed back for `openshift-priv` jobs.
